db/migrations/20191202091203_create_historical_urn_state_table_and_triggers.sql

```diff
@@ -0,0 +1,319 @@
+-- +goose Up
+CREATE TABLE api.historical_urn_state
+(
+    urn_identifier TEXT,
+    ilk_identifier TEXT,
+    block_height   BIGINT,
+    ink            NUMERIC   DEFAULT NULL,
+    art            NUMERIC   DEFAULT NULL,
+    created        TIMESTAMP DEFAULT NULL,
+    updated        TIMESTAMP NOT NULL,
+    PRIMARY KEY (urn_identifier, ilk_identifier, block_height)
+);
+
+
+CREATE FUNCTION urn_ink_before_block(urn_id INTEGER, header_id INTEGER) RETURNS NUMERIC AS
+$$
+WITH passed_block_number AS (
+    SELECT block_number
+    FROM public.headers
+    WHERE id = header_id)
+SELECT ink
+FROM maker.vat_urn_ink
+         LEFT JOIN public.headers ON vat_urn_ink.header_id = headers.id
+WHERE vat_urn_ink.urn_id = urn_ink_before_block.urn_id
+  AND headers.block_number < (SELECT block_number FROM passed_block_number)
+ORDER BY block_number DESC
+LIMIT 1
+$$
+    LANGUAGE sql;
+
+
+CREATE FUNCTION urn_art_before_block(urn_id INTEGER, header_id INTEGER) RETURNS NUMERIC AS
+$$
+WITH passed_block_number AS (
+    SELECT block_number
+    FROM public.headers
+    WHERE id = header_id)
+SELECT art
+FROM maker.vat_urn_art
+         LEFT JOIN public.headers ON vat_urn_art.header_id = headers.id
+WHERE vat_urn_art.urn_id = urn_art_before_block.urn_id
+  AND headers.block_number < (SELECT block_number FROM passed_block_number)
+ORDER BY block_number DESC
+LIMIT 1
+$$
+    LANGUAGE sql;
+
+
+CREATE FUNCTION urn_time_created(urn_id INTEGER) RETURNS TIMESTAMP AS
+$$
+SELECT api.epoch_to_datetime(MIN(block_timestamp))
+FROM maker.vat_urn_ink
+         LEFT JOIN public.headers ON vat_urn_ink.header_id = headers.id
+WHERE vat_urn_ink.urn_id = urn_time_created.urn_id
+$$
+    LANGUAGE sql;
+
+
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION maker.delete_redundant_urn_state(urn_id INTEGER, header_id INTEGER) RETURNS api.historical_urn_state
```

Comment:
Haha yeah for sure 😅 

It sounds like you got the gist of it, but it just deletes a `historical_urn_state` row if it's identical to the previous one. It's necessary because 1 row can have data from both an `ink` diff and an `art` diff, so deleting 1 kind of diff doesn't mean you can delete the associated row in this table. First we have to check that the row doesn't contain data from any other diffs, and only then can we delete it.

I agree that it's not very clear what's happening. I can at least add a comment and maybe find a clearer way to accomplish the same thing.